### PR TITLE
fix: Transfer dataSource cannot be immutable 

### LINKED ISSUE #28662: Transfer dataSource 赋值为 immutable 数据后再设置 rowKey 出错。
- [x] I have searched the [issues](https://github.com/ant-design/ant-design/issues) of this repository and believe that this is not a duplicate.

### Reproduction link
[Edit on CodeSandbox](https://codesandbox.io/s/jibenyongfa-antd4100-forked-reibp?file=/index.js)

### Steps to reproduce
1. dataSource 赋值为 immutable 数据。
2. 设置 rowKey。

### What is expected?
rowKey 被正确设置。

### What is actually happening?
控制台报错。
Cannot add property key, object is not extensible

| Environment | Info |
|---|---|
| antd | 4.10.0 |
| React | 16.14.0 |
| System | macOS 10.15.7 |
| Browser | Chrome  87.0.4280.88 |

---
1.实际场景中数据为 apollo client 取到的数据。

### PR DESCRIPTION
**Pull request by afc163**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [ ] New feature
- [x] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 🔗 Related issue link

close #28662

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | Fix  Transfer throws `Cannot add property key, object is not extensible` when `dataSource` is immutable  |
| 🇨🇳 Chinese | 修复 Transfer `dataSource` 为 immutable 数据时报 `Cannot add property key, object is not extensible` 的问题。 |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

-----
[View rendered components/transfer/demo/basic.md](https://github.com/ant-design/ant-design/blob/fix-transfer-datasource/components/transfer/demo/basic.md)

### CODE REVIEW

**Review by shaodahong — APPROVED** [review]
LGTM

File: components/transfer/index.tsx [hunk 1]
```diff
@@ -331,7 +331,10 @@ class Transfer<RecordType extends TransferItem = TransferItem> extends React.Com
     const rightDataSource: KeyWise<RecordType>[] = new Array(targetKeys.length);
     dataSource.forEach((record: KeyWise<RecordType>) => {
       if (rowKey) {
-        record.key = rowKey(record);
+        record = {
+          ...record,
+          key: rowKey(record),
+        };
```
Comment (afc163) [inline components/transfer/index.tsx:337]: 避免直接修改对象。

Comment (shaodahong) [reply components/transfer/index.tsx:337]: 嗯，这种用户传进来的修改都要 clone 一份，不然很容易出问题

Comment (theprimone) [reply components/transfer/index.tsx:337]: fork 下来跟进官方的更新还挺费劲 \_(:3J∠)\_

**Pull request merged**